modules/um_callerid.c

```diff
@@ -0,0 +1,243 @@
+/*
+ * modules/um_callerid.c
+ * Copyright (c) 2020 Ariadne Conill <ariadne@dereferenced.org>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the author may not be used to endorse or promote products
+ *    derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+ * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "stdinc.h"
+#include "modules.h"
+#include "hook.h"
+#include "client.h"
+#include "ircd.h"
+#include "send.h"
+#include "hash.h"
+#include "s_conf.h"
+#include "s_user.h"
+#include "s_serv.h"
+#include "numeric.h"
+#include "privilege.h"
+#include "s_newconf.h"
+#include "hook.h"
+#include "supported.h"
+#include "logger.h"
+
+static int
+um_callerid_modinit(void)
+{
+	user_modes['g'] = find_umode_slot();
+	if (!user_modes['g'])
+	{
+		ierror("um_callerid: unable to allocate usermode slot for +g; unloading module.");
+		return -1;
+	}
+
+	user_modes['G'] = find_umode_slot();
+	if (!user_modes['G'])
+	{
+		ierror("um_callerid: unable to allocate usermode slot for +G; unloading module.");
+		return -1;
+	}
+
+	construct_umodebuf();
+
+	add_isupport("CALLERID", isupport_string, "g");
+
+	return 0;
+}
+
+static void
+um_callerid_moddeinit(void)
+{
+	user_modes['g'] = 0;
+	user_modes['G'] = 0;
+	construct_umodebuf();
+
+	delete_isupport("CALLERID");
+}
+
+#define IsSetCallerID(c)	((c->umodes & user_modes['g']) == user_modes['g'])
+#define IsSetRelaxedCallerID(c)	((c->umodes & user_modes['G']) == user_modes['G'])
+
+static const char um_callerid_desc[] =
+	"Provides usermode +g which restricts messages from unauthorized users.";
+
+static bool
+has_common_channel(struct Client *source_p, struct Client *target_p)
```

Comment:
While I remember, this should also be returning `true` / `false`, rather than relying on the compiler to do the boolean conversion for us. It just makes it easier to read.